Template and view generation

trestrantham opened the pull request:
Creating this issue to track creation of standard scaffolding for addict templates and views.

Continuing conversation from #22:
The plan is to create an addict folder in web/templates to house these 5 templates with a shared view at web/views/addict.ex. The templates would be normal forms rather than submitting via javascript.

This approach will add `mix addict.gen.view_templates` or something similar to #9.

trenpixster commented:
As I stated previously, that's a great idea and it'll be a great base ground for implementing the other `mix` tasks :+1:

trestrantham commented:
So after looking into this further, we're going to need at least two actions for each bit of functionality we want to support if we're going to use true forms rather than POSTing via javascript. First instance, we will need a show action to render the registration form and a create action to do the POST processing (as our controller actions do currently).

In an effort to keep a RESTful approach, what are your thoughts on following the pattern [devise](https://github.com/plataformatec/devise) uses and creating multiple controllers each with corresponding `show`/`create`/etc. actions? Initially, these controllers would consist of `sessions` for logging in/out, `registrations` for registering (and possibly editing an existing account?), and `passwords` to manage/reset passwords.

If we went this route, we can just leverage our `addict :routes` helper to generate all applicable routes without burdening the user/dev. All routes would also conform to the pattern of `registrations_path(@conn, :show)` for rendering the appropriate view template and `registrations_path(@conn, :create)` for the POST paths.

Thoughts?

trenpixster commented:
Very valid points raised there.

Devise is the de facto standard indeed, and separating concerns into various controllers seems to be the architecturally correct move. The only if is that it'll make overriding default behaviour a bit more verbose with the current scheme of customization we provide though.

Devs will have to create separate controllers for each of the operations they want to customize (register, passwords and session). An alternative would be to be able to specify the controllers for the actions in the Addict configuration parameters. Any idea on different approaches for this?

jxs commented:
Hi @trestrantham, what if instead of using two actions, we define a method that returns the selected rendered that can then be overriden? 
Then define a JSON renderer controller and a HTML renderer controller that extend the base, and if the user wants to still extend more, they just extend the base  controller

trestrantham commented:
@trenpixster: Overriding the default behavior wouldn't be any different than how it is now done with the routes helper. The developer would only ever need to override the POST actions and they are easily done via `addict :routes`. We'll be generating the new/show routes via the helper as well but they wouldn't need to be overwritten from a controller perspective. Changes to the template files (`web/templates/addict/x`) directly would suffice there, right?

Going along with that :point_up: idea, it might _not_ make sense to split everything up into separate controllers. The new/show routes and actions are essentially hardcoded and having the POST actions in a single place for the developer to override is valuable.

I'll try to put something tangible together and push to WIP PR so that we can better talk through what's going on here.

@jxs: I follow at a high level what you are talking about but I'm having a hard time visualizing the actual implementation. Is there any way you could throw together a quick gist to better explain your idea?

jxs commented:
I have just read the documentation about choosing the [default render](http://www.phoenixframework.org/v0.9.0/docs/controllers) via plug.
so @trestrantham my idea is that, if we define the default render plug on the `AddictController`, and  `BaseController` agnostic , changing just the functions to return `conn`
Then define a plug that renders json instead (this should even be part of phoenix maybe) 
like [this](https://gist.github.com/jxs/bc318def237338064f00)
what do you think?

trestrantham commented:
I've added some preliminary code for this feature. There's still some hardcoded cross pollination between Addict and ExampleApp currently to get the example to work. For now, only "/register" route will work for GET requests. The rest of the routes can easily be updated once `:register` is working as expected.

One of the issues I've been having is determining where the cutoff between package (Addict) and application (ExampleApp) occurs. We know controllers will live fully in Addict (while allowing overriding action and/or controllers by the dev) and templates will live in the host application (copied over to `example_app/web/templates/addict` via generator).

I'm still unclear of where the view itself goes, though. I've added a stubbed view to the example app in this current example that just `use`s the Addict view helper, but maybe this should live in the package too?

Layouts are another issue (and are what is currently hardcoded). Although, after talking with José, the best route forward might be to add a scope macro to our routes helper that utilizes `conn.private` to pass in layout info:

```
[11:17:28]  josevalim:  addict_scope [layout: ...] do ... end
[11:17:38]  josevalim:  and you convert the scope options to private: [addict: scope_options]
...
[11:19:42]  josevalim:  just call it addict options do ... end
[11:20:24]  josevalim:  and that just expands to scope private: [addict: options] do ... end
[11:20:34]  josevalim:  and now you can access all the addict information inside conn.private.addict
```

Would love thoughts on this general direction thus far and also on the current items outlined above.

Cheers!

jxs commented:
Hi @trestrantham only saw this now, sorry. 
what i was suggesting, is what i did on the PR i submitted. Your pr seems nice to me, only have a couple issues,
1 - instead of `Addict.AddictController` imho controllers be `Addict.BaseUserController` and `Addict.UserController` what do you think @trenpixster ?
2- why are the functions on the view in the `using`macro, instead of what was before, having them defined in the main module

cheers :)

trestrantham commented:
@jxs: I commented on your PR with a :+1: and feedback on the approach.

For item 2 above, this PR changes `View` to be a `ViewHelper` and I will be adding a mix task to create an `ExampleApp.AddictView` in addition to the templates. This `AddictView` will `use` the helper and provide an entry point for a developer to further customize the views used in addict's templates.

I'll be pushing more changes up shortly, but the result of these changes will be that a mix task will copy a single view (`<AppName>.AddictView`) to `web/views` and the 5 templates to `web/templates/addict`. The routes helper will default to layout `{<AppName>.LayoutView, "application.html"}` and view `<AppName>.AddictView` but will provide options for the developer to override both via `addict :routes`:

``` elixir
addict :routes,
  register: [layout: false, view: MyApp.MyView], # layout: false renders without any layout
  login: [layout: {MyApp.MyLayout, "some_template.html"}]
```

trestrantham commented:
@trenpixster @jxs: This should be ready for review now. I'd like to incorporate changes from #25 as a followup to work done in this branch (probably in a new PR).

This isn't perfect by any means but it could use some eyes on it now.

In my mind what's next in this or separate PRs is:
1. Add tests
2. Incorporate changes outlined in #25 specifically `render_html`
3. Refactor interactors to include one per process action (login, register, etc.)--these 'top level' interactors can call other interactors as well

trestrantham commented:
Any thoughts on combining `AddictController` and `BaseController` until there's a further need to split them apart in the future?

jxs commented:
hi @trestrantham i agree on that, haven't yet looked at the rest of the PR sorry! will soon

trenpixster commented on lib/addict/controllers/addict_controller.ex:
```diff
@@ -0,0 +1,9 @@
+defmodule Addict.AddictController do
```
Can we rename this to `Addict.Controller`? I had this kind of naming before but I don't like it much :)

trenpixster commented on lib/addict/controllers/base_controller.ex:
```diff
@@ -0,0 +1,113 @@
+            {Module.concat([Utils.app_name(conn), LayoutView]), "application.html"}
+          _ ->
+            addict.layout
+        end
```
It'd be great to extract this `case` into a method to make it even more explicit.

trenpixster commented on lib/addict/utils.ex:
```diff
@@ -0,0 +1,13 @@
+defmodule Addict.Utils do
```
This is pretty cool, it's the kind of stuff Phoenix should have in the code base. PR there maybe? :smile:

trenpixster commented on lib/addict/view_helper.ex:
```diff
@@ -0,0 +1,37 @@
+      import Plug.Conn
+
+      @doc """
+      checks if user is logged in, returns true if so,
```
Can you please capitalize the `@doc`s?

trestrantham replied:
No problem :ok_hand:

trenpixster commented on lib/addict/routes_helper.ex:
```diff
@@ -12,21 +12,40 @@ defmodule Addict.RoutesHelper do
     as         = route
 
-    %{path: path, controller: controller, action: action, as: as}
+    post_action = "process_#{to_string(action)}" |> String.to_atom
```
On the ExampleApp we have to change `UserController.signout/2` to `UserController.process_signout/2`.

trestrantham replied:
I'll need to confirm this, but prefixing `signout` shouldn't be necessary. I'm doing that internally within the addict controller but that doesn't mean the custom user controllers will need to use the same naming convention since they will be specified manually in the call to `addict :routes`.

trenpixster commented on lib/tasks/templates.ex:
```diff
@@ -0,0 +1,49 @@
+  @shortdoc "Generate templates and view for Addict"
+  @moduledoc """
+  Generates an Addict view and templates
+      mix addict.gen.templates
```
Maybe instead of `templates` which colides with the Phoenix templates naming, we could call this `addict.gen.screens`.

trestrantham replied:
We prefix this mix task with `addict` to keep from any namespace collisions. I'd prefer to name it by what it does (create templates) to keep from confusion if possible but I also have no problems with `screens` if you feel strongly. :smile:

trenpixster replied:
Ah, I might've explained myself in the wrong way. The collision I refer is one of "semantics"  and not in code collision. Since we ultimately generate both `templates` and `views` calling it `templates` seems a bit weird to me :smile:

trestrantham replied:
Makes sense! I'll update to `screens`.

trenpixster commented:
Hey @trestrantham, I'm really sorry for not giving feedback earlier.

I left some comments on the code, here's some more general feedback:
- I agree that we could still have `Addict.BaseController` and `Addict.Controller` in the same file to bring context to whoever goes through the code.
- Can you update the `index.html.eex` on the ExampleApp so it links to the several new routes?
- Regarding refactoring interactors to include one per process action (login, register, etc.), totally agree with this, but it can be done in another branch, what do you think?
- It would indeed be great to have tests in this PR :+1: 

Awesome job here! 💚💛💙

jxs commented:
Controller should not be named just Controller otherwise the reverse call is not going to work

```
$ mix phoenix.routes                                                                                                                                  
 web_socket_path  GET   /phoenix           Phoenix.Transports.WebSocket.upgrade/2
 web_socket_path  POST  /phoenix           Phoenix.Transports.WebSocket.upgrade/2
 long_poller_path  GET   /phoenix/poll      Phoenix.Transports.LongPoller.poll/2
 long_poller_path  POST  /phoenix/poll      Phoenix.Transports.LongPoller.publish/2
       page_path  GET   /                  Myapp.PageController.index/2
           _path  POST  /register          Addict.Controller.register/2
           _path  POST  /login             Addict.Controller.login/2
           _path  GET   /logout            Addict.Controller.logout/2
           _path  POST  /recover_password  Addict.Controller.recover_password/2
           _path  POST  /reset_password    Addict.Controller.reset_password/2
```

trenpixster commented:
@trestrantham can you give me access to your fork so I can give you a hand with the stuff we're missing? Thanks!

trestrantham commented:
Done. Sorry for the inactivity lately! It's been a busy few weeks for me.

trenpixster commented:
Cool! No problem, same thing on my side :smile:

trenpixster commented on lib/tasks/screens.ex:
```diff
@@ -0,0 +1,49 @@
+
+    view_file = Path.join(["web", "views", "addict_view.ex"])
+                |> Path.relative_to(Mix.Project.app_path)
+    create_file view_file, view_template(bindings)
```
Hey @trestrantham, I was looking at the way this is implemented in detail and I don't quite get where this `view_template` comes from?

trestrantham replied:
@trenpixster It's a macro that's generated by calling `embed_template :view`. You can find the documentation here: http://elixir-lang.readthedocs.org/en/latest/ref/master/mix/Mix.Generator/#Mix.Generator.embed_template/2

dukex commented:
Nice work guys! What need be done to this pull request be merged?

sotte commented:
This would be a unique selling point compared to the other authentication libs. What is the state? What needs to be done (except a rebase)?

trestrantham commented:
I haven't had time in a while to work on this but please feel free to pick up where I left off. I'd start with a rebase and bumping the Elixir and Phoenix dependencies to something recent. If those go smoothly, I'd say this would largely be ready for a merge pending any further feedback from @trenpixster.

trenpixster commented:
@trestrantham I ended up doing a major refactor and implemented a version of this via #68. Thank you so much for all the hard work, I drank quite a bit from this, would love to have you take a look on how it is now :green_heart: Closing this for now!

trestrantham commented:
Your refactor looks great. Glad this made it in!

> On Apr 6, 2016, at 5:31 PM, Nizar Venturini notifications@github.com wrote:
> 
> @trestrantham I ended up doing a major refactor and implemented a version of this via #68. Thank you so much for all the hard work, I drank quite a bit from this, would love to have you take a look on how it is now  Closing this for now!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
